[BPF] Remove 'may_goto 0' instructions

yonghong-song opened the pull request:
Emil Tsalapatis from Meta reported such a case where 'may_goto 0' insn is generated by clang compiler. But 'may_goto 0' insn is actually a no-op so it makes sense to remove that in llvm. The patch is also able to handle the following code pattern
```
   ...
   may_goto 2
   may_goto 1
   may_goto 0
   ...
```
where three may_goto insns can all be removed.

Commit:
[BPF] Remove 'may_goto 0' instructions

Emil Tsalapatis from Meta reported such a case where 'may_goto 0' insn is
generated by clang compiler. But 'may_goto 0' insn is actually a no-op so
it makes sense to remove that in llvm. The patch is also able to handle
the following code pattern
   ...
   may_goto 2
   may_goto 1
   may_goto 0
   ...
where three may_goto insns can all be removed.

4ast commented:
wouldn't it be cleaner to do after inline asm was processed into proper insns  ? Instead of string match?

eddyz87 commented:
I agree with Alexei, this should be done at machine instruction level.

eddyz87 commented:
On the other hand, after assembly instructions are compiled, there is no way to distinguish `may_goto +0` from `may_goto <param>`, where param is 0. Removal of explicit `may_goto +0` goes against what we do for `goto +0`.

yonghong-song commented:
matchAndEmitInstruction() handles one instruction only. So this is not a good place to do 'may_goto 0' elimination work.

yonghong-song commented:
I checked other targets (X86, Arm) etc. They also parse and processes asm strings. For example, in ARM/ARMISelLowering.cpp,

```
ARM/ARMISelLowering.cpp:

bool ARMTargetLowering::ExpandInlineAsm(CallInst *CI) const {
  // Looking for "rev" which is V6+.
  if (!Subtarget->hasV6Ops())
    return false;

  InlineAsm *IA = cast<InlineAsm>(CI->getCalledOperand());
  StringRef AsmStr = IA->getAsmString();
  SmallVector<StringRef, 4> AsmPieces;
  SplitString(AsmStr, AsmPieces, ";\n");

  switch (AsmPieces.size()) {
  default: return false;
  case 1:
    AsmStr = AsmPieces[0];
    AsmPieces.clear();
    SplitString(AsmStr, AsmPieces, " \t,");

    // rev $0, $1
    if (AsmPieces.size() == 3 &&
        AsmPieces[0] == "rev" && AsmPieces[1] == "$0" && AsmPieces[2] == "$1" &&
        IA->getConstraintString().compare(0, 4, "=l,l") == 0) {
      IntegerType *Ty = dyn_cast<IntegerType>(CI->getType());
      if (Ty && Ty->getBitWidth() == 32)
        return IntrinsicLowering::LowerToByteSwap(CI);
    }
    break;
  }

  return false;
}
```

So use one of backend passes might be the better choice.

eddyz87 commented:
And the following example appears to miscompiled:

```c
$ cat test.c
void bar() {
  asm volatile goto ("may_goto %l[lbl];"::::lbl);
  asm volatile goto ("may_goto %l[lbl];"::::lbl);
  asm volatile goto ("may_goto %l[lbl];"::::lbl);
  asm volatile goto ("r0 = 42;");
 lbl:
  return;
}
{llvm} 12:45:19 tmp$ clang -O2 --target=bpf -S test.c -o - 2>&1 | head -n20
	.file	"test.c"
	.text
	.globl	bar                             # -- Begin function bar
	.p2align	3
	.type	bar,@function
bar:                                    # @bar
# %bb.0:                                # %entry
	#APP
	may_goto LBB0_4

	#NO_APP
# %bb.1:                                # %asm.fallthrough
	#APP
	may_goto LBB0_4

	#NO_APP
# %bb.2:                                # %asm.fallthrough1
	#APP
	may_goto LBB0_4
```

(Note the disappearance of the `"r0 = 42;"`).

eddyz87 commented on llvm/lib/Target/BPF/BPFMIPeephole.cpp:
```diff
@@ -682,6 +685,65 @@ bool BPFMIPreEmitPeephole::insertMissingCallerSavedSpills() {
+  MachineBasicBlock *ToErase = nullptr;
+
+  for (MachineBasicBlock &MBB : reverse(*MF)) {
+    if (ToErase) {
```
Note: there is a utility function `llvm::make_early_inc_range`, I think it could be used here to avoid the `ToErase` variable.

yonghong-song commented:
@eddyz87 Thanks for the error cases and suggestion to avoid ToErase variable. Will investigate all these issues.

Commit:
Check INLINEASM_BR instead of INLINEASM

Fix an issue where INLINEASM_BR asm opcode should be checked instead
of generic INLINEASM since later on retrieving target basic block
needs INLINEASM_BR.

yonghong-song commented:
Updated a new revision to address @eddyz87 reported issues. Specifically, the changes are:
  - Check INLINEASM_BR instead of INLINEASM, this is the source of errors reported by @eddyz87 earlier
  -  Use llvm::make_early_inc_range for better code

eddyz87 commented on llvm/lib/Target/BPF/BPFMIPeephole.cpp:
```diff
@@ -682,6 +685,56 @@ bool BPFMIPreEmitPeephole::insertMissingCallerSavedSpills() {
+    // The asm insn must be a may_goto insn.
+    SmallVector<StringRef, 4> AsmOpPieces;
+    SplitString(AsmPieces[0], AsmOpPieces, " ");
+    if (AsmOpPieces[0] != "may_goto")
```
I think that `AsmOpPieces[1]` should be checked as well (and total number of pieces).
Otherwise the algorithm would produce strange results for the following (incorrect) inputs:

```c
void bar() {
  asm volatile goto ("may_goto +1;"::::lbl); // should not be removed
 lbl:
  asm volatile("r0 = 42;");
  return;
}
```

```c
void bar() {
  asm volatile goto ("may_goto foobar;"::::lbl); // should not be removed, asm parser should report error later
 lbl:
  asm volatile("r0 = 42;");
  return;
}
```

yonghong-song replied:
Make sense. I did further check for asm format. That is, it must be 'may_goto $0' or 'may_goto ${0:l}'. Note that '$0'  is the *first* argument of the asm code. '${0:l}' is the first *label* argument of the asm code. Since we already enforced asm opcode INLINEASM_BR so '$0' must be refer to a label.

eddyz87 commented on llvm/lib/Target/BPF/BPFMIPeephole.cpp:
```diff
@@ -682,6 +685,56 @@ bool BPFMIPreEmitPeephole::insertMissingCallerSavedSpills() {
+
+    // Get the may_goto branch target.
+    MachineOperand &MO = MI.getOperand(InlineAsm::MIOp_FirstOperand + 1);
+    if (MO.getMBB() != Prev_MBB)
```
I think that `MO.isMBB()` is needed, otherwise this would abort on (bogus) input as below:

```
$ cat test.c
void bar(int i) {
  asm volatile goto ("may_goto %l[lbl];"::"r"(i)::lbl);
 lbl:
  asm volatile("r0 = 42;");
  return;
}
$ clang -O2 --target=bpf -S test.c -o - 2>&1 | head -n 20
clang: /home/eddy/work/llvm-project/llvm/include/llvm/CodeGen/MachineOperand.h:572: MachineBasicBlock *llvm::MachineOperand::getMBB() const: Assertion `isMBB() && "Wrong MachineOperand accessor"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: clang -O2 --target=bpf -S test.c -o -
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module 'test.c'.
4.	Running pass 'BPF PreEmit Peephole Optimization' on function '@bar'
 #0 0x00007f8711bfe388 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/eddy/work/llvm-project/llvm/lib/Support/Unix/Signals.inc:794:13
 #1 0x00007f8711bfc260 llvm::sys::RunSignalHandlers() /home/eddy/work/llvm-project/llvm/lib/Support/Signals.cpp:106:18
 #2 0x00007f8711b3e7d6 (anonymous namespace)::CrashRecoveryContextImpl::HandleCrash(int, unsigned long) /home/eddy/work/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:73:5
 #3 0x00007f8711b3e7d6 CrashRecoverySignalHandler(int) /home/eddy/work/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:390:51
 #4 0x00007f8711426090 __restore_rt (/lib64/libc.so.6+0x1a090)
 #5 0x00007f871147f0f4 __pthread_kill_implementation /usr/src/debug/glibc-2.40-17.fc41.x86_64/nptl/pthread_kill.c:44:76
 #6 0x00007f8711425fde gsignal /usr/src/debug/glibc-2.40-17.fc41.x86_64/signal/../sysdeps/posix/raise.c:27:6
 #7 0x00007f871140d942 abort /usr/src/debug/glibc-2.40-17.fc41.x86_64/stdlib/abort.c:81:7
 #8 0x00007f871140d85e _nl_load_domain.cold /usr/src/debug/glibc-2.40-17.fc41.x86_64/intl/loadmsgcat.c:1177:9
 #9 0x00007f871141e107 (/lib64/libc.so.6+0x12107)
#10 0x00007f8718304657 llvm::MachineOperand::getSymbolName() const /home/eddy/work/llvm-project/llvm/include/llvm/CodeGen/MachineOperand.h:638:5
#11 0x00007f8718304657 (anonymous namespace)::BPFMIPreEmitPeephole::removeMayGotoZero() /home/eddy/work/llvm-project/llvm/lib/Target/BPF/BPFMIPeephole.cpp:702:43
```

yonghong-song replied:
The update enforced the asm format, so the new code should avoid the above situation.

eddyz87 replied:
I think `.isMBB()` check is still needed, as the bogus input below triggers `Assertion `isMBB() && "Wrong MachineOperand accessor"'`:

```c
void bar(int i) {
  asm volatile goto ("may_goto %[i];"::[i]"r"(i)::lbl);
 lbl:
  asm volatile("r0 = 42;");
  return;
}
```

yonghong-song replied:
You are correct. Thanks!

eddyz87 commented on llvm/lib/Target/BPF/BPFMIPeephole.cpp:
```diff
@@ -682,6 +685,56 @@ bool BPFMIPreEmitPeephole::insertMissingCallerSavedSpills() {
+
+    const char *AsmStr = MI.getOperand(0).getSymbolName();
+    SmallVector<StringRef, 4> AsmPieces;
+    SplitString(AsmStr, AsmPieces, ";\n");
```
Maybe count `";\n"` instead of splitting?
To avoid allocations for unrelated assembly blocks.

yonghong-song replied:
I used the approach similar to ARM/ARMISelLowering.cpp:
```
  InlineAsm *IA = cast<InlineAsm>(CI->getCalledOperand());
  StringRef AsmStr = IA->getAsmString();
  SmallVector<StringRef, 4> AsmPieces;
  SplitString(AsmStr, AsmPieces, ";\n");

  switch (AsmPieces.size()) {
  default: return false;
  case 1:
    AsmStr = AsmPieces[0];
    AsmPieces.clear();
    SplitString(AsmStr, AsmPieces, " \t,");

    // rev $0, $1
    if (AsmPieces.size() == 3 &&
        AsmPieces[0] == "rev" && AsmPieces[1] == "$0" && AsmPieces[2] == "$1" &&
        IA->getConstraintString().compare(0, 4, "=l,l") == 0) {
      IntegerType *Ty = dyn_cast<IntegerType>(CI->getType());
      if (Ty && Ty->getBitWidth() == 32)
        return IntrinsicLowering::LowerToByteSwap(CI);
    }
    break;
  }
```
The code already enforced only one asm insn is allowed. see
```
    // Do not support multiple insns in one inline asm.
    if (AsmPieces.size() != 1)
      continue;
```

Maybe I missed something here?

eddyz87 replied:
I was thinking about something like below to avoid needless allocations. Unfortunately, there is no version of `StringRef::count` that searches for a bag of characters, so it all looks a bit clumsy, feel free to ignore if you don't like it.

```diff
diff --git a/llvm/lib/Target/BPF/BPFMIPeephole.cpp b/llvm/lib/Target/BPF/BPFMIPeephole.cpp
index 5f1891d3fce2..d86dc799cc21 100644
--- a/llvm/lib/Target/BPF/BPFMIPeephole.cpp
+++ b/llvm/lib/Target/BPF/BPFMIPeephole.cpp
@@ -685,6 +685,15 @@ bool BPFMIPreEmitPeephole::insertMissingCallerSavedSpills() {
   return Changed;
 }
 
+static StringRef isSingleAsmInsn(StringRef Str) {
+  size_t P = Str.find_last_of(";\n");
+  if (P == StringRef::npos)
+    return Str;
+  if (P == Str.size() - 1)
+    return Str.substr(0, P);
+  return {};
+}
+
 bool BPFMIPreEmitPeephole::removeMayGotoZero() {
   bool Changed = false;
   MachineBasicBlock *Prev_MBB, *Curr_MBB = nullptr;
@@ -699,17 +708,15 @@ bool BPFMIPreEmitPeephole::removeMayGotoZero() {
     if (MI.getOpcode() != TargetOpcode::INLINEASM_BR)
       continue;
 
-    const char *AsmStr = MI.getOperand(0).getSymbolName();
-    SmallVector<StringRef, 4> AsmPieces;
-    SplitString(AsmStr, AsmPieces, ";\n");
-
+    StringRef AsmStr = MI.getOperand(0).getSymbolName();
     // Do not support multiple insns in one inline asm.
-    if (AsmPieces.size() != 1)
+    AsmStr = isSingleAsmInsn(AsmStr);
+    if (AsmStr.empty() || !AsmStr.starts_with("may_goto"))
       continue;
 
     // The asm insn must be a may_goto insn.
     SmallVector<StringRef, 4> AsmOpPieces;
-    SplitString(AsmPieces[0], AsmOpPieces, " ");
+    SplitString(AsmStr, AsmOpPieces, " ");
     if (AsmOpPieces.size() != 2 || AsmOpPieces[0] != "may_goto")
       continue;
     // Enforce the format of 'may_goto <label>'.
```

eddyz87 commented on llvm/lib/Target/BPF/BPFMIPeephole.cpp:
```diff
@@ -682,6 +685,56 @@ bool BPFMIPreEmitPeephole::insertMissingCallerSavedSpills() {
   return Changed;
 }
 
+bool BPFMIPreEmitPeephole::removeMayGotoZero() {
```
When I build BPF selftests with this patch I get the following error:
[cpuv4.error.log](https://github.com/user-attachments/files/18542254/cpuv4.error.log)

yonghong-song replied:
With the newer fix, I checked with bpf selftests, all tests passed.

eddyz87 replied:
I tried building `test_progs-cpuv4` and get the following error (note, llvm is compiled with `-DLLVM_ENABLE_ASSERTIONS=ON`):

```
clang  -g -Wall -Werror -D__TARGET_ARCH_x86 -mlittle-endian -I/home/eddy/work/bpf-next/tools/testing/selftests/bpf/tools/include -I/home/eddy/work/bpf-next/tools/testing/selftests/bpf -I/home/eddy/work/bpf-next/tools/include/uapi -I/home/eddy/work/bpf-next/tools/testing/selftests/usr/include -std=gnu11 -fno-strict-aliasing -Wno-compare-distinct-pointer-types -idirafter /home/eddy/work/llvm-project/build/lib/clang/20/include -idirafter /usr/local/include -idirafter /usr/include    -DENABLE_ATOMICS_TESTS   -O2 --target=bpfel -c progs/bpf_flow.c -mcpu=v4 -o /home/eddy/work/bpf-next/tools/testing/selftests/bpf/cpuv4/bpf_flow.bpf.o
clang: /home/eddy/work/llvm-project/llvm/include/llvm/ADT/ilist_iterator.h:168: reference llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::MachineInstr, true, true, void, false, void>, false, false>::operator*() const [OptionsT = llvm::ilist_detail::node_options<llvm::MachineInstr, true, true, void, false, void>, IsReverse = false, IsConst = false]: Assertion `!NodePtr->isKnownSentinel()' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: clang -g -Wall -Werror -D__TARGET_ARCH_x86 -mlittle-endian -I/home/eddy/work/bpf-next/tools/testing/selftests/bpf/tools/include -I/home/eddy/work/bpf-next/tools/testing/selftests/bpf -I/home/eddy/work/bpf-next/tools/include/uapi -I/home/eddy/work/bpf-next/tools/testing/selftests/usr/include -std=gnu11 -fno-strict-aliasing -Wno-compare-distinct-pointer-types -idirafter /home/eddy/work/llvm-project/build/lib/clang/20/include -idirafter /usr/local/include -idirafter /usr/include -DENABLE_ATOMICS_TESTS -O2 --target=bpfel -c progs/bpf_flow.c -mcpu=v4 -o /home/eddy/work/bpf-next/tools/testing/selftests/bpf/cpuv4/bpf_flow.bpf.o
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module 'progs/bpf_flow.c'.
4.	Running pass 'BPF PreEmit Peephole Optimization' on function '@flow_dissector_1'
 #0 0x00007f38a6dfefb8 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/eddy/work/llvm-project/llvm/lib/Support/Unix/Signals.inc:798:13
 #1 0x00007f38a6dfce90 llvm::sys::RunSignalHandlers() /home/eddy/work/llvm-project/llvm/lib/Support/Signals.cpp:106:18
 #2 0x00007f38a6d3f3d6 (anonymous namespace)::CrashRecoveryContextImpl::HandleCrash(int, unsigned long) /home/eddy/work/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:73:5
 #3 0x00007f38a6d3f3d6 CrashRecoverySignalHandler(int) /home/eddy/work/llvm-project/llvm/lib/Support/CrashRecoveryContext.cpp:390:51
 #4 0x00007f38a6626090 __restore_rt (/lib64/libc.so.6+0x1a090)
 #5 0x00007f38a667f0f4 __pthread_kill_implementation /usr/src/debug/glibc-2.40-17.fc41.x86_64/nptl/pthread_kill.c:44:76
 #6 0x00007f38a6625fde gsignal /usr/src/debug/glibc-2.40-17.fc41.x86_64/signal/../sysdeps/posix/raise.c:27:6
 #7 0x00007f38a660d942 abort /usr/src/debug/glibc-2.40-17.fc41.x86_64/stdlib/abort.c:81:7
 #8 0x00007f38a660d85e _nl_load_domain.cold /usr/src/debug/glibc-2.40-17.fc41.x86_64/intl/loadmsgcat.c:1177:9
 #9 0x00007f38a661e107 (/lib64/libc.so.6+0x12107)
#10 0x00007f38ad57a8f4 llvm::ilist_detail::node_base_prevnext<llvm::ilist_node_base<true, void>, true>::isSentinel() const /home/eddy/work/llvm-project/llvm/include/llvm/ADT/ilist_node_base.h:44:36
#11 0x00007f38ad57a8f4 llvm::ilist_detail::node_base_prevnext<llvm::ilist_node_base<true, void>, true>::isKnownSentinel() const /home/eddy/work/llvm-project/llvm/include/llvm/ADT/ilist_node_base.h:45:41
#12 0x00007f38ad57a8f4 llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::MachineInstr, true, true, void, false, void>, false, false>::operator*() const /home/eddy/work/llvm-project/llvm/include/llvm/ADT/ilist_iterator.h:168:5
#13 0x00007f38ad57a8f4 llvm::MachineInstrBundleIterator<llvm::MachineInstr, false>::operator*() const /home/eddy/work/llvm-project/llvm/include/llvm/CodeGen/MachineInstrBundleIterator.h:178:40
#14 0x00007f38ad57a8f4 llvm::MachineBasicBlock::back() /home/eddy/work/llvm-project/llvm/include/llvm/CodeGen/MachineBasicBlock.h:335:52
#15 0x00007f38ad5794fe (anonymous namespace)::BPFMIPreEmitPeephole::removeMayGotoZero() /home/eddy/work/llvm-project/llvm/lib/Target/BPF/BPFMIPeephole.cpp:699:24
#16 0x00007f38ad5794fe (anonymous namespace)::BPFMIPreEmitPeephole::runOnMachineFunction(llvm::MachineFunction&) /home/eddy/work/llvm-project/llvm/lib/Target/BPF/BPFMIPeephole.cpp:342:16
```

yonghong-song replied:
Indeed, enabling -DLLVM_ENABLE_ASSERTIONS=ON can expose the issue. The reason is due to one basic block is empty and we need to check that.

I just updated a new version to fix this.

Commit:
Ensure correct may_goto asm format

Check AsmOpPieces[1] to ensure may_goto asm format correct.

Commit:
Check may_goto operand must be a BasicBlock and MBB must not be empty
